Build out Gatsby media information in docs (#12906)
gatsbyjs/gatsby: docs/docs/working-with-video.md

import Video from "../components/video"

export default function VlogTemplate({
  data, // this prop will be injected by the GraphQL query below.
}) {
  const { markdownRemark } = data // data.markdownRemark holds our post data
  const { frontmatter, html } = markdownRemark
  return (
    <div className="blog-post-container">
      <div className="blog-post">
        <h1>{frontmatter.title}</h1>
        <h2>{frontmatter.date}</h2>
        <Video
          videoSrcURL={frontmatter.videoSrcURL}
          videoTitle={frontmatter.videoTitle}
        />
        <div
          className="blog-post-content"
          dangerouslySetInnerHTML={{ __html: html }}
        />
      </div>
    </div>
  )
}

export const pageQuery = graphql`
  query($path: String!) {
    markdownRemark(frontmatter: { path: { eq: $path } }) {
      html
      frontmatter {
        date(formatString: "MMMM DD, YYYY")
        path
        title
        videoSrcURL
        videoTitle
      }
    }
  }
`
```

## Hosting your own HTML5 video files

It's super common to source video from YouTube, Twitch or Vimeo. But what if you want to host your own video and include it as HTML5 video?

> > This discussion is ongoing on Github, chime in with your ideas in the Gatsby issue [#3346 Create a special component for HTML5 videos](https://github.com/gatsbyjs/gatsby/issues/3346)

Reviewer: Branding nitpick
```suggestion
> This discussion is ongoing on GitHub, chime in with your ideas in the Gatsby issue [#3346 Create a special component for HTML5 videos](https://github.com/gatsbyjs/gatsby/issues/3346)
```

To include your own video files that will work in multiple web browsers and platforms, you'll need to read up a bit on video extensions and codecs. We recommend MDN as a resource: [Media formats for HTML audio and video](https://developer.mozilla.org/en-US/docs/Web/HTML/Supported_media_formats). You may need video converter software to produce the necessary formats to support a range of devices and environments, such as `.webm` and `.mp4`.

## Using custom video players

One advantage of integrating a custom component with your own hosted video is it can give you more control over the video player, including its accessibility. It is strongly encouraged to provide captions and subtitles for your videos, and use a player with accessible controls.

> Check out the accessible [HTML5 video player from Paypal](https://github.com/paypal/accessible-html5-video-player#react-version) for an example compatible with Gatsby and React.

Reviewer: Nitpick for branding.
```suggestion
Check out the accessible [HTML5 video player from PayPal](https://github.com/paypal/accessible-html5-video-player#react-version) for an example compatible with Gatsby and React.
```

---

> title: Working With Video*

Reviewer: I don't think the asterisk should be included here — it should only be in the sidebar YML to trigger the "stub" styling. (I think.)

Reviewer 2: This is correct! Titles stay clean, the yaml gets updated with an asterisk!

issue: https://github.com/gatsbyjs/gatsby/issues/12903
---

> This is a stub. Help our community expand it.

Reviewer: Is this really a stub? This has a _ton_ of information!

Please use the [Gatsby Style Guide](/contributing/gatsby-style-guide/) to ensure your
pull request gets accepted.

> The discussion around best practices for video in Gatsby is also ongoing on Github, chime in with your ideas here: https://github.com/gatsbyjs/gatsby/issues/3346

Reviewer 3: "talking" Links

```suggestion
The discussion around best practices for video in Gatsby is also ongoing on Github, chime in with your ideas in the Gatsby issue [#3346 Create a special component for HTML5 videos](https://github.com/gatsbyjs/gatsby/issues/3346)
```

---

> 

Reviewer 3: I miss a heading here like TOC or `Overview`. But I don't find the right wording now

Author: I'm not sure what you mean. The heading for the page is right above it (Working with Video). We've talked about creating more consistent patterns for TOC, such as in #6896. But I don't think that should hold up this PR!

Reviewer 3: I did mean a header for the TOC. 

But I did not find any doc with a header for TOC. Only some introducing sentence on some pages following by TOC or links to aubpages.

- [Sourcing video from a host](#sourcing-video-from-a-host)
- [Embedding hosted videos in Markdown](#embedding-hosted-videos-in-markdown)
- [Writing custom components for hosted video](#writing-custom-components-for-hosted-video)
- [Querying video data from Markdown with GraphQL](#querying-video-data-from-markdown-with-graphql)
- [Hosting your own HTML5 video files](#hosting-your-own-html5-video-files)
- [Using custom video players](#using-custom-video-players)

## Sourcing video from a host

The easiest method for including video on a Gatsby site is to source an uploaded file from a site like YouTube, Vimeo, or Twitch. Using the source URL from one of those hosts, you can use Remark plugins or create a custom `<iframe>` solution to embed videos into your Gatsby site.

## Embedding hosted videos in Markdown

> There are numerous Gatsby plugins for working with hosted video in your Markdown posts and pages. We recommend checking out the [gatsby-remark-embed-video](<(/packages/gatsby-remark-embed-video/?=video)>) plugin for sourcing from a variety of hosts.

Reviewer: This link looks broken (see the `(<(`). I think some weird wrapping happened?

Author: Oh yeah, looks like something weird with my editor. Will fix!

```jsx:title=src/pages/404.js
import React from "react"

import Layout from "../components/layout"
import SEO from "../components/seo"
import Video from "../components/video"

const NotFoundPage = () => (
  <Layout>
    <SEO title="404: Not found" />
    <section>
      <h1>NOT FOUND</h1>
      <p>You just hit a page that doesn't exist... the sadness.</p>
      <p>May I suggest a video instead?</p>
      <Video
        videoSrcURL="https://www.youtube.com/embed/dQw4w9WgXcQ"
        videoTitle="Official Music Video on YouTube"
      />
    </section>
  </Layout>
)

export default NotFoundPage
```

## Querying video data from Markdown with GraphQL

If a Markdown page or post has a featured video, you might want to include a video URL and title in [its frontmatter](/docs/adding-markdown-pages#note-on-creating-markdown-files). That makes it easy to pass those values into our custom component:

```markdown:title=my-first-post.md
---
path: "/blog/my-first-post"
date: "2019-03-27"
title: "My first blog post"
videoSourceURL: https://www.youtube.com/embed/dQw4w9WgXcQ

> videoTitle: "Official Music Video on YouTube"

Reviewer 2: ```suggestion
videoTitle: "Gatsby is Never Gonna Give You Up"
```

😆

import Video from "../components/video"

export default function VlogTemplate({
  data, // this prop will be injected by the GraphQL query below.
}) {
  const { markdownRemark } = data // data.markdownRemark holds our post data
  const { frontmatter, html } = markdownRemark
  return (
    <div className="blog-post-container">
      <div className="blog-post">
        <h1>{frontmatter.title}</h1>
        <h2>{frontmatter.date}</h2>
        <Video
          videoSrcURL={frontmatter.videoSrcURL}
          videoTitle={frontmatter.videoTitle}
        />
        <div
          className="blog-post-content"
          dangerouslySetInnerHTML={{ __html: html }}
        />
      </div>
    </div>
  )
}

export const pageQuery = graphql`
  query($path: String!) {
    markdownRemark(frontmatter: { path: { eq: $path } }) {
      html
      frontmatter {
        date(formatString: "MMMM DD, YYYY")
        path
        title
        videoSrcURL
        videoTitle
      }
    }
  }
`
```

## Hosting your own HTML5 video files

It's super common to source video from YouTube, Twitch or Vimeo. But what if you want to host your own video and include it as HTML5 video?

> > This discussion is ongoing on Github, chime in with your ideas here: https://github.com/gatsbyjs/gatsby/issues/3346

Reviewer 3: ```suggestion
> This discussion is ongoing on Github, chime in with your ideas in the Gatsby issue [#3346 Create a special component for HTML5 videos](https://github.com/gatsbyjs/gatsby/issues/3346)
```

> To include your own video files that will work in multiple web browsers and platforms, you'll need to read up a bit on video extensions and codecs. We recommend MDN as a resource: <a href="https://developer.mozilla.org/en-US/docs/Web/HTML/Supported_media_formats">Media formats for HTML audio and video</a>. You may need video converter software to produce the necessary formats to support a range of devices and environments, such as `.webm` and `.mp4`.

Reviewer: Not a big deal, but why not use Markdown for this link?

## Using custom video players

One advantage of integrating a custom component with your own hosted video is it can give you more control over the video player, including its accessibility. It is strongly encouraged to provide captions and subtitles for your videos, and use a player with accessible controls.

> Check out the accessible HTML5 video player for React from Paypal: https://github.com/paypal/accessible-html5-video-player#react-version

Reviewer: Should this be linkified?

Author: I thought Markdown would convert that automatically, but I will check! Thanks for the tip.

Reviewer: It will — I should have clarified. I meant turning it into a named link, e.g. `"PayPal has created an [accessible HTML5 video player](#link)."`

---

> This is a stub. Help our community expand it.

Reviewer: Looks like this is doubled up.
